core/src/main/java/org/apache/calcite/rel/RelShuttleImpl.java

```diff
@@ -44,24 +42,18 @@
  * any children change.
  */
 public class RelShuttleImpl implements RelShuttle {
-  protected final Deque<RelNode> stack = new ArrayDeque<>();
 
   /**
    * Visits a particular child of a parent.
    */
   protected RelNode visitChild(RelNode parent, int i, RelNode child) {
-    stack.push(parent);
-    try {
```

Comment:
The stack only keep references to the Nodes objects that would never be garbage collected by Java during the recursive visit pattern  invoking. How can it be an overhead.

Reply:
Stack needs to be created, resized and populated. Considering content is never read, I consider this overhead (especially when queries are deep trees).